plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/tasks/GenerateReportsTask.kt

```diff
@@ -0,0 +1,174 @@
+package org.jlleitschuh.gradle.ktlint.tasks
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.file.ProjectLayout
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.SetProperty
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Classpath
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.PathSensitive
+import org.gradle.api.tasks.PathSensitivity
+import org.gradle.api.tasks.TaskAction
+import org.gradle.workers.WorkerExecutor
+import org.jetbrains.kotlin.konan.file.File
+import org.jlleitschuh.gradle.ktlint.reporter.ReporterType
+import org.jlleitschuh.gradle.ktlint.worker.ConsoleReportWorkAction
+import org.jlleitschuh.gradle.ktlint.worker.GenerateReportsWorkAction
+import org.jlleitschuh.gradle.ktlint.worker.LoadReportersWorkAction
+import java.io.FileInputStream
+import java.io.ObjectInputStream
+import javax.inject.Inject
+
+/**
+ * Generates reports and prints errors into Gradle console.
+ *
+ * This will actually fail the build in case some non-corrected lint issues.
+ */
+@CacheableTask
+abstract class GenerateReportsTask @Inject constructor(
+    private val workerExecutor: WorkerExecutor,
+    projectLayout: ProjectLayout,
+    objectFactory: ObjectFactory
+) : DefaultTask() {
```

Comment:
Sorry, I'm not a huge fan of having this be a separate task. This pattern is in conflict with how most tasks in Gradle operate.

Pretty much all Gradle tasks both do the work and generate the report.

That way, if you invoke the specific task manually from the task selector, you always get the report.

Won't this break that?

Reply:
It was intentional to separate actual linting and generating reports into separate tasks:
- allows in the future to split linting into file buckets that will be handled by workers that run in parallel. This should improve plugin tasks execution time on single module projects with a lot of source files.
- KtLint reports should be generated in one place, so this will prevent splitting linting in buckets.
- Later allow to print all errors from all `SourceSet`s into console in final `ktlintCheck` or `ktlintFormat` tasks
- Probably it will be possible to implement this in the same task, but currently Gradle does not allow to order workers (I am not sure if it is even a good idea to support), so output of one worker treated as input of another one.

> That way, if you invoke the specific task manually from the task selector, you always get the report.

You are wrong here:
- if no input `SourceSet`s are not changed, generate report task will be in `UP_TO_DATE` status
- if input `SourceSet` are changed, generate reports task will still invoke first lint task as generate reports task inputs depends on lint task output.

Reply:
Not sure this would fit the use cases you want to implement but you could consider something like `lintTask.finalizedBy(corresondingReportTask)` so that when one invoke a linting task directly, its companion report task is run afterwards.

Reply:
@Tapchicoma please do do this `finalizedBy`